src/bin/ion/commands/beta/to/json.rs

```diff
@@ -57,75 +41,44 @@ pub fn run(_command_name: &str, matches: &ArgMatches) -> Result<()> {
     };
 
-    if let Some(input_file_iter) = matches.get_many::<String>("input") {
-        for input_file in input_file_iter {
-            let mut file = File::open(input_file)
-                .with_context(|| format!("Could not open file '{}'", input_file))?;
-            convert(&mut file, &mut output, format)?;
+    if let Some(input_file_names) = matches.get_many::<String>("input") {
+        // Input files were specified, run the converter on each of them in turn
+        for input_file in input_file_names {
+            let file = File::open(input_file.as_str())
+                .with_context(|| format!("Could not open file '{}'", &input_file))?;
+            let mut reader = ReaderBuilder::new()
+                .build(file)
+                .with_context(|| format!("Input file {} was not valid Ion.", &input_file))?;
+            convert(&mut reader, &mut output)?;
         }
     } else {
-        // If no input file was specified, run the inspector on STDIN.
-
-        // The inspector expects its input to be a byte array or mmap()ed file acting as a byte
-        // array. If the user wishes to provide data on STDIN, we'll need to copy those bytes to
-        // a temporary file and then read from that.
-
-        // Create a temporary file that will delete itself when the program ends.
-        let mut input_file = tempfile::tempfile().with_context(|| {
-            concat!(
-                "Failed to create a temporary file to store STDIN.",
-                "Try passing an --input flag instead."
-            )
-        })?;
-
-        // Pipe the data from STDIN to the temporary file.
-        let mut writer = BufWriter::new(input_file);
-        io::copy(&mut io::stdin(), &mut writer)
-            .with_context(|| "Failed to copy STDIN to a temp file.")?;
-        // Get our file handle back from the BufWriter
-        input_file = writer
-            .into_inner()
-            .with_context(|| "Failed to read from temp file containing STDIN data.")?;
-        convert(&mut input_file, &mut output, format)?;
+        // No input files were specified, run the converter on STDIN.
+        let mut reader = ReaderBuilder::new()
+            .build(stdin().lock())
+            .with_context(|| "Input was not valid Ion.")?;
+        convert(&mut reader, &mut output)?;
     }
 
     output.flush()?;
     Ok(())
 }
 
-pub fn convert(file: &mut File, output: &mut Box<dyn Write>, format: &str) -> Result<()> {
-    // NOTE: mmap logic is copied from inspect.inspect_file().
-
-    // mmap involves operating system interactions that inherently place its usage outside of Rust's
-    // safety guarantees. If the file is unexpectedly truncated while it's being read, for example,
-    // problems could arise.
-    let mmap = unsafe {
-        MmapOptions::new()
-            .map(file)
-            .with_context(|| "Could not mmap ")?
-    };
-
-    // Treat the mmap as a byte array.
-    let ion_data: &[u8] = &mmap[..];
```

Comment:
🗺️ `ion-rs` now supports iterating over the elements in a buffered reader, so we no longer have to `mmap` the input file. This also means that we can read directly from `STDIN` without first dumping it to a temporary file.

Reply:
🎉